File: 13. Roman to Integer/romanToInt.go
package main

import "fmt"

func romanToInt(s string) int {
	result := 0
	for i, c := range s {
		switch c {
		case 'I':
			result++
		case 'V':
			if i > 0 && s[i-1] == 'I' {
				result += 3
			} else {
				result += 5
			}
		case 'X':
			if i > 0 && s[i-1] == 'I' {
				result += 8
			} else {
				result += 10
			}
		case 'L':
			if i > 0 && s[i-1] == 'X' {
				result += 30
			} else {
				result += 50
			}
		case 'C':
			if i > 0 && s[i-1] == 'X' {
				result += 80
			} else {
				result += 100
			}
		case 'D':
			if i > 0 && s[i-1] == 'C' {
				result += 300
			} else {
				result += 500
			}
		case 'M':
			if i > 0 && s[i-1] == 'C' {
				result += 800
			} else {
				result += 1000
			}
		}
	}
	return result
}

func main() {
	fmt.Println(romanToInt("III"))
	fmt.Println(romanToInt("IV"))
	fmt.Println(romanToInt("IX"))
	fmt.Println(romanToInt("LVIII"))
	fmt.Println(romanToInt("MCMXCIV"))
}

File: 13. Roman to Integer/romanToInt.py
def romanToInt(s: str) -> int:
    result = 0
    dict = {
        "I": 1,
        "IV": 4,
        "V": 5,
        "IX": 9,
        "X": 10,
        "XL": 40,
        "L": 50,
        "XC": 90,
        "C": 100,
        "CD": 400,
        "D": 500,
        "CM": 900,
        "M": 1000,
    }
    while len(s) > 0:
        if s[0:2] in dict:
            result += dict[s[0:2]]
            s = s[2::]
        else:
            result += dict[s[0:1]]
            s = s[1::]

    return result


def romanToInt2(s: str) -> int:
    result = 0
    dict = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
    temp = 0
    for i in reversed(s):
        if dict[i] >= temp:
            result = result + dict[i]
        else:
            result = result - dict[i]
        temp = dict[i]

    return result


print(romanToInt2("IV"))
print("ANS: 4")
print(romanToInt2("IX"))
print("ANS: 9")
print(romanToInt2("LVIII"))
print("ANS: 58")
print(romanToInt2("MCMXCIV"))
print("ANS: 1994")
